### PR TITLE
Better notebook markup renderer api

### PR DESCRIPTION
**Pull request by mjbvz**
For #121256

- Use js modules for notebook output renderers
- Rename apis from `markdown` to `markup`
- Use imports and exports for apis instead of globals for apis
- Use esbuild instead of webpack so we can emit js modules
- Clearly split top level markup renderers from renderers that extend other renderers

### CODE REVIEW

**Comment by mjbvz** [comment]
/cc @rebornix and @connor4312.

For the API, the main two files to look at are:

- `extensions/markdown-language-features/notebook/index.ts` — This defines a top level markup renderer. We are using it here to render markdown cells with markdown-it
 
- `extensions/notebook-markdown-extensions/notebook/katex.ts` — This shows how another renderer can then extend a markup renderer. This idea is that the top level renderer controls how these dependencies are pulled int

The api still needs to be discussed and refined, but I like that it lets me use JS modules and exports for the core API. May be things to consider for the output renderers api

File: build/gulpfile.extensions.js [hunk 1]
```diff
@@ -336,4 +345,47 @@ async function webpackExtensions(taskName, isWatch, webpackConfigLocations) {
...
+async function esbuildExtensions(taskName, isWatch, outputRoot, esbuildScripts) {
+	function reporter(/** @type {string} */ stdError, /** @type {string} */script) {
+		const matches = (stdError || '').match(/\> (.+): error: (.+)?/g);
+		fancyLog(`Finished ${ansiColors.green(taskName)} ${script} with ${matches ? matches.length : 0} errors.`);
+		for (const match of matches || []) {
+			fancyLog.error(match);
+		}
+	}
+
+	const scripts = esbuildScripts.map(script => {
```
Comment (connor4312) [inline build/gulpfile.extensions.js:363]: esbuild has a pretty good JavaScript API they they actually recommend using for more complicated builds. Could consider that to reduce complexity and child process overhead here.

Comment (connor4312) [reply build/gulpfile.extensions.js:363]: Oh, I see you use that below. I think `esbuild.build` will spawn a child process with the native esbuild binary, so nesting a child process may be unnecessary

File: src/vs/workbench/contrib/notebook/browser/extensionPoint.ts [hunk 2]
```diff
@@ -42,16 +42,20 @@ export interface INotebookRendererContribution {
...
 
-export interface INotebookMarkdownRendererContribution {
-	readonly [NotebookMarkdownRendererContribution.id]?: string;
-	readonly [NotebookMarkdownRendererContribution.displayName]: string;
-	readonly [NotebookMarkdownRendererContribution.entrypoint]: string;
+export interface INotebookMarkupRendererContribution {
+	readonly [NotebookMarkupRendererContribution.id]?: string;
+	readonly [NotebookMarkupRendererContribution.displayName]: string;
+	readonly [NotebookMarkupRendererContribution.entrypoint]: string;
+	readonly [NotebookMarkupRendererContribution.dependsOn]: string | undefined;
```
Comment (connor4312) [inline src/vs/workbench/contrib/notebook/browser/extensionPoint.ts:57]: In generic output renderers we call this `dependencies`, we should consider doing so here for consistency (even if its length will be <= 1)

File: src/vs/workbench/contrib/notebook/browser/view/renderers/webviewPreloads.ts [hunk 3]
```diff
@@ -1002,11 +1001,24 @@ function webviewPreloads() {
...
+	const markdownCtx = {
+		dependencies: markdownRenderer.dependencies,
+	};
+
+	return `${webviewPreloads}`
+		.slice(0, -1)
+		.replace(/function webviewPreloads\(\) \{/g, '')
+		.replace(/__outputNodePadding__/g, `${styleValues.outputNodePadding}`)
+		.replace(/__outputNodeLeftPadding__/g, `${styleValues.outputNodeLeftPadding}`)
+		.replace(/\/\*__markdownRendererBlock__\*\//g, `
```
Comment (connor4312) [inline src/vs/workbench/contrib/notebook/browser/view/renderers/webviewPreloads.ts:1020]: Does the comment survive minification in production builds?

**Pull request merged**